### PR TITLE
balloon_hotplug.with-migrate

### PR DESCRIPTION
**Pull request by ybduan**
balloon_hotplug.with-migrate: Add migration test after hotplugging/unplugging balloon device

ID: 1467186

Signed-off-by: yduan yduan@redhat.com

### CODE REVIEW

**Comment by ybduan** [comment]
(1/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug: PASS (159.09 s)
 (2/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug.with_reboot: PASS (344.87 s)
 (3/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug.with_system_reset: PASS (351.50 s)
 (4/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug.with_shutdown_after_unplug: PASS (161.07 s)
 (5/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug.with_shutdown_after_plug: PASS (148.02 s)
 (6/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug.with_migrate.migrate_after_hotplug: PASS (203.37 s)
 (7/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug.with_migrate.migrate_after_unplug: PASS (195.49 s)

**Comment by jingzhao84** [comment]
@ybduan how can you make sure balloon device unplug successfully before migration (Line107)

Thanks
Jing

**Comment by ybduan** [comment]
There is a function named "verify_unplug" at line 92. @jingzhao84

**Comment by huangyum** [comment]
I suggest to verify guest dmesg after test. Other parts look good to me. Thanks.

**Comment by ybduan** [comment]
(1/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug: PASS (1333.29 s)
 (2/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug.with_reboot: PASS (341.91 s)
 (3/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug.with_system_reset: PASS (351.99 s)
 (4/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug.with_shutdown_after_unplug: PASS (157.35 s)
 (5/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug.with_shutdown_after_plug: PASS (148.43 s)
 (6/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug.with_migrate.migrate_after_hotplug: PASS (208.94 s)
 (7/7) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.balloon_hotplug.with_migrate.migrate_after_unplug: PASS (197.06 s)

**Comment by ybduan** [comment]
Add function verify_kernel_crash() after test, please help to review again, thanks! @jingzhao84 @huangyum

**Comment by jingzhao84** [comment]
@huangyum  could you help to review this pr, thanks

**Comment by huangyum** [comment]
LGTM,
Acked-by: Yumei Huang <yuhuang@redhat.com>

**Comment by jingzhao84** [comment]
@luckyh  could you help to review this pr, thanks

**Comment by luckyh** [comment]
@lijinlijin pls help review this patch, thanks.

File: qemu/tests/balloon_hotplug.py [hunk 1]
```diff
@@ -66,6 +65,10 @@ def run_pm_test(pm_test, plug_type):
             err += ("\nDevice is not in qtree %ss after hotplug:\n%s"
                     % (pause, vm.monitor.info("qtree")))
 
+        vm.params["balloon"] = "balloon%d" % idx
+        vm.params["balloon_dev_devid"] = "balloon%d" % idx
```
Comment (lijinlijin) [inline qemu/tests/balloon_hotplug.py]: seems idx is always 0 according to line42, could we replace idx with 0 directly?

**Comment by lijinlijin** [comment]
job passed on win10-64 as well: 
(1/2) smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.balloon_hotplug.with_migrate.migrate_after_hotplug: PASS (409.51 s)
 (2/2) smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.balloon_hotplug.with_migrate.migrate_after_unplug: PASS (362.04 s)

File: qemu/tests/cfg/balloon_hotplug.cfg [hunk 2]
```diff
@@ -35,3 +35,12 @@
         - with_shutdown_after_plug:
             pm_test_after_plug = shutdown
             check_img = yes
+        - with_migrate:
+            test_tags = "evict"
```
Comment (lijinlijin) [inline qemu/tests/cfg/balloon_hotplug.cfg]: Is there any reason why test only evict in migrate cases?
for other cases, we tested both enlarge and evict:
 - test_tags = "evict enlarge"

File: qemu/tests/cfg/balloon_hotplug.cfg [hunk 3]
```diff
@@ -35,3 +35,12 @@
...
+            variants:
+                - migrate_after_hotplug:
+                    run_sub_test_after_balloon = yes
+                    sub_test_after_balloon = migration
+                    migration_test_command = help
```
Comment (lijinlijin) [inline qemu/tests/cfg/balloon_hotplug.cfg]: This line could be removed.

File: qemu/tests/balloon_hotplug.py (hunk 1 above)
Comment (ybduan) [reply qemu/tests/balloon_hotplug.py]: Hard coding is not a good choice, thanks!

File: qemu/tests/cfg/balloon_hotplug.cfg (hunk 2 above)
Comment (ybduan) [reply qemu/tests/cfg/balloon_hotplug.cfg]: Modified, thanks!

File: qemu/tests/cfg/balloon_hotplug.cfg (hunk 3 above)
Comment (ybduan) [reply qemu/tests/cfg/balloon_hotplug.cfg]: Modified, thanks!

File: qemu/tests/balloon_hotplug.py [hunk 4]
```diff
@@ -92,11 +95,22 @@ def run_pm_test(pm_test, plug_type):
...
+                del vm.params["balloon_dev_add_bus"]
+                vm.migrate(float(params.get("mig_timeout", "3600")))
+
             if pm_test_after_unplug:
                 run_pm_test(pm_test_after_unplug, "hot-unplug")
                 if not vm.is_alive():
                     return
+
+    error_context.context("Check kernel crash message!", logging.info)
+    vm.verify_kernel_crash
```
Comment (luckyh) [inline qemu/tests/balloon_hotplug.py]: Typo.

File: qemu/tests/balloon_hotplug.py [hunk 5]
```diff
@@ -66,6 +65,10 @@ def run_pm_test(pm_test, plug_type):
             err += ("\nDevice is not in qtree %ss after hotplug:\n%s"
                     % (pause, vm.monitor.info("qtree")))
 
+        vm.params["balloon"] = "balloon%d" % idx
+        vm.params["balloon_dev_devid"] = "balloon%d" % idx
+        vm.params["balloon_dev_add_bus"] = "yes"
```
Comment (luckyh) [inline qemu/tests/balloon_hotplug.py]: Why add these params here and delete them at line 105?

File: qemu/tests/balloon_hotplug.py [hunk 6]
```diff
@@ -92,11 +95,22 @@ def run_pm_test(pm_test, plug_type):
...
+                del vm.params["balloon_dev_devid"]
+                del vm.params["balloon_dev_add_bus"]
+                vm.migrate(float(params.get("mig_timeout", "3600")))
+
             if pm_test_after_unplug:
                 run_pm_test(pm_test_after_unplug, "hot-unplug")
                 if not vm.is_alive():
                     return
+
+    error_context.context("Check kernel crash message!", logging.info)
```
Comment (luckyh) [inline qemu/tests/balloon_hotplug.py]: Suggest `"Verify guest alive"`.

Comment (ybduan) [reply qemu/tests/balloon_hotplug.py]: Modified, thanks!

File: qemu/tests/balloon_hotplug.py (hunk 5 above)
Comment (ybduan) [reply qemu/tests/balloon_hotplug.py]: Do adding after hotplugging and deleting after hotunplugging. Thanks!

File: qemu/tests/balloon_hotplug.py (hunk 4 above)
Comment (ybduan) [reply qemu/tests/balloon_hotplug.py]: Modified, thanks!

**Comment by luckyh** [comment]
Thanks all.

**Pull request merged**